### PR TITLE
@safe class comparisons

### PR DESCRIPTION
**Pull request by adamdruppe**
I laid out in the ProtoObject dip thread how the real problem has nothing to do with Object, but rather with a helper function. It unnecessarily goes through Object instead of using the types actually passed to it and thus misses the static information available to it.

By simply making the helper function into a template, it is able to use the full static information and thus infer its attributes.

There is a migration path here though: users will want to get away from Object.opEquals. Delete the word `override` and you can offer an overload instead, which can then more specified parameters. We might want to outright deprecate Object.opEquals to inform people about the migration path, but even with it still being there, you can add the overload today, as I did in TypeInfo (both as an example and because the top-level opEquals calls it, so we wouldn't want to .

The downside of this migration - if you want to call it a downside - is if you wanted to cast both to Object and still use ==, you're in trouble. Even if you override the generic method and do the ugly casts to forward to your good function, the overloading rules will issue an ambiguity error. (Which I think it shouldn't - the more derived one is obviously a better choice, and that's how overloading works in other cases, so why not here? Yes, I know, because that's how the spec defines the partial ordering. But I still don't like it. Nevertheless, it is what we have.)

So you have to just not override the Object.opEquals, meaning if you compare through Object, you will always get the base Object behavior (`a is b`). But if it is deprecated anyway, you'd just stop doing that.

Now, all that said, nothing about this new code actually breaks the old thing if you do override it, you just don't get the new benefits. With one exception: it wouldn't work with cast, since I deleted the overload that casted it away. To maintain compatibility with that, I added the static if check in a second commit. I really dislike that - I'd prefer it just be a compile error - but since it is a legacy compatibility issue, this hack should keep things rolling.

Should we decide to actually deprecate and remove the Object methods, then this hack can also be eliminated.

The unittests are modified to show what is now `@safe` - the overload is respect - and what must remain `@system` - using the legacy Object comparison.

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

**Comment by adamdruppe** [comment]
Of course the AA implementation uses the legacy function.... I guess I have no choice but to offer some way to override it while overloading for the new capabilities. See the next commit for how.

This will surely require a documentation update as well. It is always a bit messy to do this kind of thing but hopefully it will appease the auto tester.

**Comment by adamdruppe** [comment]
These tests are so fragile..... the failing ones are dmd assuming very specific details about Object.

dmd and druntime absolutely must be a monorepo. They obviously are closely intertwined.

**Comment by adamdruppe** [comment]
Of course the other option to maintain compatibility is to make the helper function check for the presence of opEqualsEx and if it is present, use it, otherwise fall back on the opEquals, which would then become the legacy function. I don't like that; I'd rather keep the canonical name as opEquals, just overload the signature.

File: src/object.d [hunk 1]
```diff
@@ -237,41 +242,42 @@ class Object
     }
 }
 
-bool opEquals(Object lhs, Object rhs)
+bool opEquals(LHS, RHS)(LHS lhs, RHS rhs)
```
[the remark this answers is not in the document]
Comment (adamdruppe) [reply src/object.d:240]: pretty sure that's a specialization not a restriction... a constraint could always be done if need be tho

**Comment by MoonlightSentinel** [comment]
Btw druntime could provide a `mixin` that inserts the necessary boilerplate to enable virtual dispatch:

Rough first draft:
```d
/// Overrides opEquals by calling into custom method accepting typeof(this)
mixin template forwardingOpEquals()
{
    // TODO: Check that another opEquals exists
    override bool opEquals(const Object o) const
    {
        auto a = cast(typeof(this)) o;
        return a && this.opEquals(a);
    }
}
```

User classes could then enjoy the benefits of this patch without loosing the generic equality check. E.g.
```d
class A {
    int i;
	
    this(int i)
    {
        this.i = i;
    }
    
    mixin forwardingOpEquals!();
    
    final bool opEquals(const A a) const /* applicable attributes */
    {
        return this.i == a.i;
    }
}
```

**Comment by adamdruppe** [comment]
On Fri, Jan 07, 2022 at 05:43:04PM -0800, Florian wrote:
> Btw druntime could provide a `mixin` that inserts the necessary boilerplate to enable virtual dispatch:

Yes, indeed.

**Comment by adamdruppe** [comment]
There we go, I think I got the implementation details worked out. Will come back tomorrow for a style pass and to collapse the commits and update the docs.

**Comment by adamdruppe** [comment]
I put a note in there about @nogc typeinfo.name actually blocking this from being nogc in general. That hack is there because of dmd on Windows' dll implementation. And it is specific to all those words: dmd, Windows, dll, implementation. Fix any one of them and it is no longer necessary.

But, that said, since it is type info anyway the names are statically known. We could probably put them in a temporary buffer, or CTFE calculate it into fully static data, or even change the name == name bit to be something more akin to a range or sink piece-by-piece comparison. All of those would would enable the nogc to spread upwards through this again.

**Comment by thewilsonator** [comment]
Could you please add a changelog entry for this.

**Comment by MoonlightSentinel** [comment]
> I put a note in there about @nogc typeinfo.name actually blocking this from being nogc in general. That hack is there because of dmd on Windows' dll implementation. And it is specific to all those words: dmd, Windows, dll, implementation. Fix any one of them and it is no longer necessary.

The `TypeInfo.name` check could probably be hidden behind `version (DigitalMars) version (Windows)` in a followup PR (to avoid unnecessary code)

**Comment by adamdruppe** [comment]
On Sun, Jan 09, 2022 at 02:44:32PM -0800, Nicholas Wilson wrote:
> Could you please add a changelog entry for this.

oh yeah, i forgot about that, will come back to it later tonight.

**Comment by adamdruppe** [comment]
On Sun, Jan 09, 2022 at 02:50:32PM -0800, Florian wrote:
> The `TypeInfo.name` check could probably be hidden behind `version (DigitalMars) version (Windows)` in a followup PR (to avoid unnecessary code)

Oh possibly, but that'd lead to people possibly saying X compiles with
nogc in ldc but not in dmd and that mightbe iffy.

**Comment by adamdruppe** [comment]
small changelog entry but it is there now and all still passing.

File: changelog/safe_opequals.dd [hunk 2]
```diff
@@ -0,0 +1,3 @@
+Add support for `@safe` class opEquals
```
Comment (RazvanN7) [inline changelog/safe_opequals.dd:1]:
```suggestion
Add support for `@safe` class `opEquals`
```

File: src/object.d [hunk 3]
```diff
@@ -180,7 +180,30 @@ class Object
     /**
      * Test whether $(D this) is equal to $(D o).
      * The default implementation only compares by identity (using the $(D is) operator).
-     * Generally, overrides for $(D opEquals) should attempt to compare objects by their contents.
+     * Generally, overrides and overloads for $(D opEquals) should attempt to compare objects by their contents.
+     * You will most likely want to add an overload that takes your specific type as the argument
```
Comment (RazvanN7) [inline src/object.d:184]: I don't mean to ponder upon petty details, but this is public documentation so it is best if we avoid having a casual tone.
"You will most likely want to add an overload ..." does not have the neutral, professional tone that is expected from technical prose. I would rephrase this paragraph as to get rid of the uses of "you". Something along the lines of: "The most common case is represented by overloads that take a specific type as argument ...".

File: src/object.d [hunk 4]
```diff
@@ -180,7 +180,30 @@ class Object
...
+     * and does the content comparison. Then you can override this and forward it to your specific
```
Comment (RazvanN7) [inline src/object.d:185]: ditto.

File: src/object.d [hunk 5]
```diff
@@ -180,7 +180,30 @@ class Object
...
+     * typed overload with a cast. Remember to check for `null` on the typed overload.
```
Comment (RazvanN7) [inline src/object.d:186]:
```suggestion
     * typed overload with a cast. Null must be checked for by the user on the typed overload
```

**Review by RazvanN7 — APPROVED** [review]
LGTM minus the documentation nits.

File: src/object.d (hunk 3 above)
Comment (maxhaton) [reply src/object.d:184]: This is fine. The documentation isn't a textbook, and in fact some of the best textbooks are the ones that *do* take mostly conversational tone.

https://llvm.org/docs/CodingStandards.html contains almost 70 "you"s, for example and does it's job as documentation just fine despite actually being aimed at compiler engineers in their ivory towers.

What I will say, however, is that we should probably strive for listing a few (not just one) representative scenarios rather than going for "you will most likely". Rather than squabbling over the writing *style*, we need to ensure we are actually transmitting useful information. Well, squabble a bit, but the things that really matter are not the writing style when it comes to the documentation in this area of the *library* (the language specification is another matter).

File: changelog/safe_opequals.dd [hunk 6]
```diff
@@ -0,0 +1,3 @@
...
+
+object.opEquals is now @safe if you provide a opEquals in your class that is also @safe.
```
[the remark this answers is not in the document]
Comment (adamdruppe) [reply changelog/safe_opequals.dd]: added

File: src/object.d (hunk 3 above)
Comment (adamdruppe) [reply src/object.d]: I slightly reworded it but that's not my writing style and I'm not going to pretend it is.

If you prefer, I can just delete the documentation.

**Comment by quickfur** [comment]
@adamdruppe  Do you have a link to the dip discussion thread you referred to? I can't seem to find it anywhere.

**Comment by adamdruppe** [comment]
I also changed the title since it isn't actually nothrow in the current implementation - need a bit more typeinfo overhaul for that. It isn't terribly difficult to do but I'd prefer to keep that separated out.

**Comment by adamdruppe** [comment]
@quickfur yeah https://github.com/dlang/DIPs/pull/219#issuecomment-954740989

**Comment by RazvanN7** [comment]
Thanks @adamdruppe
